### ISSUE TITLE
Flutter support for Apple Watch

### ISSUE THREAD

**Issue by vijayredme** [flutter/flutter#28901]
is it possible to create a watchOS application with flutter ?

**Comment by Gorniv** [comment]
wait https://github.com/flutter/flutter/issues/15288

**Comment by jmagman** [comment]
Once https://github.com/flutter/flutter/issues/15288 lands, documentation will be needed for how to add Flutter to a watch app.

**Comment by ilkkanisula** [comment]
#15288 is closed now. Is there some basic instructions available so I can test to build simple flutter app for watchOS simulator?

**Comment by peazz** [comment]
bump

**Comment by zeucxb** [comment]
Is it already possible?

**Comment by zeucxb** [comment]
news?

**Comment by peazz** [comment]
You can write a native apple watch app and you can communicate with flutter via streams, you do have to link up the app yourself but it can be done.

**Comment by sebastianpicklum** [comment]
But that’s not a Flutter app on the Watch. 

Flutter should run on an Apple Watch as AFAIK everything it needs to run on iOS is available also in watchOS. 

I’ve already developed an application for watchOS with SwiftUI and I was quite impressed on how well it performs without draining the battery. So I think Flutter will also perform quite well on watchOS as it has many similarities with SwiftUI.

**Comment by nelsonic** [comment]
@peazz can please you share a bit more detail on how you would _link_ a Flutter iOS App with a Watch OS App (written in Swift) using streams? Do you know of any examples in the App Store where the devs have done this so we can test out the UX of both? (Thanks!)

Obviously this thread is for adding Apple Watch support to Flutter and that will be _amazing_, 
but until that happens, we need to understand if the intermediate step is workable.

**Comment by sebastianpicklum** [comment]
It’s done the same way a native iPhone app is linked with a Watch app. You just use streams to route payloads from/to Flutter.

**Comment by kf6gpe** [comment]
I _think_ this likely needs some documentation help from both the framework and tool teams --- assigning appropriate labels. Feel free to re-label if I'm wrong..

**Comment by jmagman** [comment]
To support a real Flutter watch app, there's more than documentation work to be done.  The engine would need to create a version of the Flutter framework APIs that links on and subclasses WatchKit or SwiftUI instead of UIKit, similar to the `FlutterMacOS.framework` version that links on AppKit for macOS. 

Figuring out bitcode was blocking _that_ work, I believe.  @dnfield Does that all sound right?

**Comment by dnfield** [comment]
Yes - the bitcode piece should not be a blocker anymore, but there still needs to be embedding code written for  WatchKit.

AFAIK, this is not on anyone's radar actively.  But it should be possible to do.

**Comment by jmagman** [comment]
I'm going to re-purpose this issue for that embedding work, then.

**Comment by jmagman** [comment]
Though when that's done where will be additional work for infra, tool, and documentation.

**Comment by kuprel** [comment]
Would complications be supported?

**Comment by sebastianpicklum** [comment]
> Would complications be supported?

I would recommend to use a platform channel to support complications. Complications work different from how Flutter work. There could be some wrapper package for Flutter, but in general, I would simply implement complications the native way - Which is really simple.

**Comment by nimit199** [comment]
Hey, is there a way to fetch live health data from iwatch using flutter?

**Comment by ilkkanisula** [comment]
You can try https://github.com/metaflowltd/flutter_health_fit

**Comment by AndroidNils** [comment]
Any updats?

**Comment by jmagman** [comment]
To use Flutter for a watch app, we would still need to write a new embedder that links on WatchKit, uses its app lifecycle, etc.  You can add a non-Flutter watch extension to your Flutter app, which would communicate over platform channels.  
All of this was previously mentioned above, so no updates.

**Comment by rdev-software** [comment]
Is there any update on wear/watchos progress ?

**Comment by jmagman** [comment]
@rdev-software 
> To use Flutter for a watch app, we would still need to write a new embedder that links on WatchKit, uses its app lifecycle, etc. You can add a non-Flutter watch extension to your Flutter app, which can communicate over platform channels.
> All of this was previously mentioned above, so no updates.

**Comment by jmagman** [comment]
> After adding watch Extention in ios/Runner.xcworkspace flutter run fails with below error
> 
> Steps to reproduce
> 
> 1. `flutter create watch_test`
> 2. `cd watch_test` & `flutter run` - App successfully launches in iOS simulator
> 3. Now open ios/Runner.xcworkspace and add new target as watch app extension for iOS app
> 4. `flutter run` will give error
> 
> * Tried on both stable & master branch
> * WatchOS targets Supported Platforms is watchOS & Build Active Arch only YES
> 
> <img alt="Screen Shot 2021-10-21 at 4 53 27 PM" width="765" src="https://user-images.githubusercontent.com/574396/138292064-3b3a2330-ab42-4e02-91f9-a5d005ec4338.png">
> 
> ')
> Error log note: Using new build system note: Planning note: Build preparation complete note: Building targets in parallel error: unable to resolve product type 'com.apple.product-type.watchkit2-extension' for platform 'iphonesimulator' (in target 'watch WatchKit Extension' from project 'Runner') error: unable to resolve product type 'com.apple.product-type.watchkit2-extension' for platform 'iphonesimulator' (in target 'watch WatchKit Extension' from project 'Runner') error: unable to resolve product type 'com.apple.product-type.application.watchapp2' for platform 'iphonesimulator' (in target 'watch' from project 'Runner') error: unable to resolve product type 'com.apple.product-type.application.watchapp2' for platform 'iphonesimulator' (in target 'watch' from project 'Runner') warning: duplicate output file '' on task: RegisterExecutionPolicyException /Users/raeeskattali/Downloads/w_test2/build/ios/Debug-iphonesimulator (in target 'watch' from project 'Runner

Please file a new issue for what you're seeing when adding a watch extension. This issue is tracking a feature request for Flutter to run on watchOS.

**Comment by otopba** [comment]
Is there any update on watchos progress ?

**Comment by diegogarciar** [comment]
Hi! I have a question: Why do we have to set bitcode when adding Apple Watch app?
I have been developing a watch app with no problem. Haven't distributed though.

**Comment by jmagman** [comment]
> Hi! I have a question: Why do we have to set bitcode when adding Apple Watch app? I have been developing a watch app with no problem. Haven't distributed though.

Apple will reject your app if `ENABLE_BITCODE=NO` for the watch target, or if you are linking against a framework that does not have bitcode enabled.

**Comment by yura2000** [comment]
Hi! Has anyone had this error **CoreTelephony is not available when building for watchOS Simulator.** when trying to build an app with Apple Watch extension? The app already exists with a lot of packages and plugins, but seems the problem isn't with it because I have imported all packages and plugins into another existing app and it works fine.

![image](https://user-images.githubusercontent.com/11773084/149473906-d8bcd442-ea46-43ea-9d5b-46657f88fcb1.png)

**Comment by ricardoboss** [comment]
How about 6 months later? Any update on this issue?

**Comment by jmagman** [comment]
> To use Flutter for a watch app, we would still need to write a new embedder that links on WatchKit, uses its app lifecycle, etc. You can add a non-Flutter watch extension to your Flutter app, which can communicate over platform channels.
All of this was previously mentioned above, so no updates.

**Comment by ciriousjoker** [comment]
Any news? P6 seems to be a bit low given the general shift towards smartwatches, especially given Google's efforts with the Pixel Watch lately.

**Comment by jmagman** [comment]
There is no update.  When there is an update someone will likely comment on this issue.  Please just 👍 to be subscribed to notifications instead of adding noise.

**Comment by gilaraujo** [comment]
> Apple will reject your app if ENABLE_BITCODE=NO for the watch target, or if you are linking against a framework that does not have bitcode enabled.

[Now that bitcode is no longer required for watchOS applications](https://developer.apple.com/documentation/Xcode-Release-Notes/xcode-14-release-notes), is there any other requirement that prevents flutter from supporting WatchOS?

Thanks in advance!

**Comment by pedromassango** [comment]
Hello @jmagman thanks for the overview on what needs to be done. My question is is the Flutter team even considering working on this?

Because there are some custom engines that we can currently use to run Flutter on apple Tv, so it seems the that the challenge is not really making it possible but instead how the Flutter team would maintain the support for Apple TV.

One experience I had was, I was able to build/run Flutter on Apple TV, and the only problem that made me discontinue the project was because Plugins (shared_preferences specifically) wasn't work and I didn't have deep knowledge of the Flutter engine to do the necessary changes.

**Comment by Rexios80** [comment]
@pedromassango
Specifically mentions Apple Watch: https://github.com/flutter/flutter/wiki/Roadmap#non-goals
Relevant: https://github.com/flutter/flutter/wiki/Popular-issues#apple-carplay--android-auto-support-26801

**Comment by jmagman** [comment]
> One experience I had was, I was able to build/run Flutter on Apple TV, and the only problem that made me discontinue the project was because Plugins (shared_preferences specifically) wasn't work and I didn't have deep knowledge of the Flutter engine to do the necessary changes.

Plugins are a pain point for custom embedders. For example the eLinux embedder made their own plugins: https://github.com/sony/flutter-elinux-plugins

Example camera plugin: 
https://github.com/sony/flutter-elinux-plugins/blob/584bb07788a2ee2472b8331c97d2c717d9149cf2/packages/camera/pubspec.yaml#L20-L21

Tooling code that parses it: https://github.com/sony/flutter-elinux/blob/51a1d685901f79fbac51665a967c3a1a789ecee5/lib/elinux_plugins.dart#L76

There's been a little work supporting sharing more code between macOS and iOS plugins (https://github.com/flutter/flutter/pull/115337) so in theory a plugin could support Apple TV and watchOS as well in a shared iOS/macOS directory, but Flutter tooling doesn't support that right now.

**Comment by pranil-iottive** [comment]
I can build the functionality with the Apple Watch app & seems to be working fine while in debug mode.
but the issue occurred with apple watch  while trying to release the app.
1. [StackOverflow Issue](https://stackoverflow.com/questions/77741752/flutter-ios-companion-app-with-watch-app-error-no-such-module-watchkit-while?noredirect=1#comment137340997_77741752)
2. If few times we perform clean build from XCode / flutter clean then getting **'Flutter/Flutter.h' file not found**

**Comment by jmagman** [comment]
> I can build the functionality with the Apple Watch app & seems to be working fine while in debug mode.
> but the issue occurred with apple watch while trying to release the app.

This issue is tracking Flutter apps on the Watch.  Are you describing a Flutter iOS app with a companion non-Flutter Watch app?  If so, that's something different, can you file a new issue with exact reproduction steps?  Attaching your Xcode project instead of just screenshots would be helpful.

**Comment by NischalJoshi777** [comment]
Any update on this guys ?

**Comment by andrejvujic** [comment]
Still no update?

**Comment by Alejandroem** [comment]
Is there an update on this?
